### ISSUE TITLE
OutOfMemoryError encountered when reading from Kafka

### ISSUE THREAD

**Issue by dan-tr** [openzipkin/zipkin#1332]
I have a test that mimics web service interactions that result in 17 spans. I can scale the test by looping (i.e. run the same test 10 times sequentially) and also by threads (i.e. use two threads to run the test two times concurrently). I can then scale the test by both of these parameters - 10 threads by 10 loops per thread results in 1700 spans (10 x 10 x 17). These spans are sent to a Kafka queue to be picked up by Zipkin-Server.
The test runs without error at 200 threads and 10 loops per thread (34000) spans, but when I try 500 threads and 10 loops (85000) spans, I get this OOM Error:

```
zipkin          | Exception in thread "pool-2-thread-1" java.lang.OutOfMemoryError: Java heap space
zipkin          |       at java.nio.HeapByteBuffer.<init>(HeapByteBuffer.java:57)
zipkin          | Exception: java.lang.OutOfMemoryError thrown from the UncaughtExceptionHandler in thread "start KafkaCollector-SendThread(zkbold-sandbox.int.thomsonreuters.com:2181)"
zipkin          |
zipkin          |       at java.nio.ByteBuffer.allocate(ByteBuffer.java:335)
zipkin          |       at sun.nio.cs.StreamDecoder.<init>(StreamDecoder.java:251)
zipkin          |       at sun.nio.cs.StreamDecoder.<init>(StreamDecoder.java:231)
zipkin          |       at sun.nio.cs.StreamDecoder.forInputStreamReader(StreamDecoder.java:69)
zipkin          |       at java.io.InputStreamReader.<init>(InputStreamReader.java:74)
zipkin          |       at zipkin.internal.JsonCodec.jsonReader(JsonCodec.java:602)
zipkin          |       at zipkin.internal.JsonCodec.readList(JsonCodec.java:582)
zipkin          |       at zipkin.internal.JsonCodec.readSpans(JsonCodec.java:432)
zipkin          |       at zipkin.collector.Collector.acceptSpans(Collector.java:93)
zipkin          |       at zipkin.collector.kafka.KafkaStreamProcessor.run(KafkaStreamProcessor.java:60)
zipkin          |       at zipkin.collector.kafka.KafkaCollector$LazyStreams$1.run(KafkaCollector.java:206)
zipkin          |       at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1142)
zipkin          |       at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:617)
zipkin          |       at java.lang.Thread.run(Thread.java:745)
```

Since Zipkin-Server is reading the spans from a Kafka queue I wouldn't think the number of messages on the queue should impact the memory consumption of the server.

**Comment by codefromthecrypt** [comment]
can you send the result of http://your_host/metrics and /dump when
that happens? I'm wondering if there's a backlog of storage operations
(consuming memory). we probably need to throttle a bit.
https://github.com/openzipkin/zipkin/issues/1181

**Comment by codefromthecrypt** [comment]
By the way, while it is possible to overrun the cluster using the native transport, I found that if using http, it doesn't.

Ex. if I use `ES_HOSTS=http://localhost:9200` (on latest zipkin), backlogging even 100K messages doesn't result in any memory or overload I can reproduce.

Can you try to make a reproducible script, such as below, so that myself or someone else can help debug this? By scripting, we don't have to guess so much. Please include the parameters you are using to run the server.

https://github.com/openzipkin/zipkin/issues/1141#issuecomment-228514983

**Comment by codefromthecrypt** [comment]
and by "overrun the cluster" I mean it is possible for us to send so many messages that we exceed the worker queue in elasticsearch, resulting in errors like this:

```
org.elasticsearch.common.util.concurrent.EsRejectedExecutionException: rejected execution of org.elasticsearch.transport.TransportService$4@6ab048e0 on EsThreadPoolExecutor[index, queue capacity = 200, org.elasticsearch.common.util.concurrent.EsThreadPoolExecutor@1cb4322a[Running, pool size = 8, active threads = 8, queued tasks = 200, completed tasks = 243]]
```

**Comment by Logic-32** [comment]
@adriancole, it looks like I need to revive this case.  We are currently running zipkin 1.28.0 and using Elasticsearch as our storage component using HTTP all the way from Java/Brave down through zipkin to Elasticsearch (so no Kafka) and are having this issue.

We have a couple of heap dumps which seem to implicate OkHttp3 as the thing causing the leaks.  There are, specifically, a bunch of okhttp3.RealCall$AsyncCall's dangling about.

Please let me know what other information I can provide to help diagnose this because we have frequent OOM issues as a result of whatever is causing this; the two most recent ones correlating with issues on Elasticsearch's side.

Also, can you provide a script for what you did to attempt and reproduce the OOM issue?  I have an idea of how I'd like to try and get something reproducible over here and it'd be nice to see if my idea is the same as yours before wasting time duplicating the effort ;)

**Comment by codefromthecrypt** [comment]
>
> @adriancole <https://github.com/adriancole>, it looks like I need to
> revive this case. We are currently running zipkin 1.28.0 and using
> Elasticsearch as our storage component using HTTP all the way from
> Java/Brave down through zipkin to Elasticsearch (so no Kafka) and are
> having this issue.
>
OOM is too generic to try to address in one issue about kafka. I did add a
note about exceptions here, which is probably why you added details about
elasticsearch. Can you describe your issue on the backpressure issue? #1741
The problem is that when ES is backlogged, requests don't complete. We may
be able to tactically put a request timeout on okhttp side so that it
doesn't linger when the server is not being nice. Another option is to rate
limit, but the other issue is more specific to ES, so probably best to
carry-forward there. Thanks for the offer to run tests!

**Comment by codefromthecrypt** [comment]
Especially for those using ES, use latest brave and zipkin and have brave
use v2 reporter against /api/v2

This will reduce gc pressure on both sides as ES is v2 native and using v1
endpoints means converting every span

**Comment by Logic-32** [comment]
Good news is, we're not using Kafka, so I suppose this is a bit off topic.  I'm happy to create a new case if so, or just comment on #1741.  However, that even seems a bit inappropriate since I'm not in a position to "talk about what is possible, especially what is easy to support."

As it stands, I'm also not sure "backpressure" is entirely the issue we're having.  The second instance we had an OOM, for example, was because Elasticsearch ran out of disk space.  Something obviously easy to remedy, but in that case I wouldn't expect zipkin-server to lock up and die like it did.  Instead, imo, it should try to push what data it gets and drop whatever Elasticsearch says it can't store; assuming Elasticsearch has the ability to say it can't store it.  To some extent, I'm curious why the ES_HTTP_READ_TIMEOUT property doesn't resolve this issue.  I'm guessing it's because OkHttp3 may not be able to establish the connection to Elasticsearch and therefore isn't getting to a point where it's waiting on a read?

With all that said, aside from the obvious reasons, would you still recommend updating to v2 to test this issue or do you suppose we'll still have the same problem but with a bit more buffer until we blow the heap?  Note, we have locally made some changes to zipkin-server that would likely have the same effect on GC/heap usage.  We were planning on doing a pull request but v2 came out before we could get some solid tests in ;)

p.s. For what it's worth, we have instrumented zipkin-server with metrics we report to Graphite and our heap remains sub-1.5GB, averaging about 500MB, until ES dies.  At which point there is a rapid, borderline immediate spike to 3GB where it then churns on GC indefinitely.  The heap dumps I mentioned earlier were taken at this point.

**Comment by codefromthecrypt** [comment]
>
> As it stands, I'm also not sure "backpressure" is entirely the issue we're
> having. The second instance we had an OOM, for example, was because
> Elasticsearch ran out of disk space. Something obviously easy to remedy,
> but in that case I wouldn't expect zipkin-server to lock up and die like it
> did. Instead, imo, it should try to push what data it gets and drop
> whatever Elasticsearch says it can't store; assuming Elasticsearch has the
> ability to say it can't store it. To some extent, I'm curious why the
> ES_HTTP_READ_TIMEOUT property doesn't resolve this issue. I'm guessing it's
> because OkHttp3 may not be able to establish the connection to
> Elasticsearch and therefore isn't getting to a point where it's waiting on
> a read?
>
I think this is as you say not really a backpressure so much as how to keep
zipkin from crashing when ES is not happy. However the solution might be
related. I would still upgrade even if it yeah won't help from crashing
when ES it dead (maybe only slow it).. pretty sure we updated OkHttp as
well lately. I've been thinking about maybe maxRequests=64 could be
problematic depending on how big the in-flight requests are. I've also
wondered if we shouldn't drop sooner, perhaps with a custom http listener
(undertow or similar) so that we don't buffer any data when the backend is
dead. Maybe we can retitle https://github.com/openzipkin/zipkin/issues/1741
towards ES resilience or open one on that point. Any help you can give
would be appreciated.. apologies the work in progress on v2 made that a bit
of shifting sands.

>

**Comment by Logic-32** [comment]
I can definitely say we occasionally throw some large requests at this.  We actually had to up the 10k span limit in ThriftCodec to 20k because we were semi-regularly getting 13k spans per message (rouge service we've identified and quieted down a bit).  On average I'd say we're in the 10s per message but with the number of services we have we get hundreds of thousands of spans a second (IIRC).  So, in short, the messages probably aren't that big on average but there are _a lot_ of them.

I'd definitely be pro dropping early if you know the backend is down though I'm not in a position to try and implement something like that myself.  We can definitely try setting 'maxRequests=something' but, out of curiosity, where would we set that?  Spring-boot, OkHttp3, StorageComponent, environment variable, other...?

No worries about moving the ground out from beneath us!  From what I can tell the changes we have will move nicely to Zipkin 2 :)  As for us actually moving to it: it'll be awhile.  But I'll try to get some momentum in that direction and report back if progress is ever made and anything positive happens.

**Comment by codefromthecrypt** [comment]
Are you sending this data via a homegrown or existing library? I know census has support for pruning spans as for example in last workshop there was no tool that could handle over 10k.. many drop far before that, for example using statistics. We have not explored this deeply as usually very large traces are an error.

Ps okhttp does have max connections but i dont think that limits backlog. We can verify.

**Comment by Logic-32** [comment]
We are using Brave with a URLConnectionSender to send the data.  I'm surprised that no tool could handle more than 10k spans in your workshop; we handle it with minimal problems on only one zipkin-server, no cluster.  The only problem is that the ES_READ_TIMEOUT is set to 60s and it sometimes takes more than 60s to write the data out.  So, the data does get written but zipkin erroneously reports them as dropped.  But our specific situation that got us into that state was having a 10s messageTimeout and a tight loop that hit the database; both of which have been remedied.

Quick review of the documentation indicates that the max connections only affects the size of the pool they use; which probably means everything else would be queued and backlogged.

Again, not knowing OkHttp that well, I wonder if you can inspect the Dispatcher queue and cancel anything in the queue if you exceed a certain threshold?
Ref: https://gist.github.com/RyanRamchandar/64c5863838940ec67f03

**Comment by codefromthecrypt** [comment]
>
> Quick review of the documentation indicates that the max connections only
> affects the size of the pool they use; which probably means everything else
> would be queued and backlogged.
>
> Again, not knowing OkHttp that well, I wonder if you can inspect the
> Dispatcher queue and cancel anything in the queue if you exceed a certain
> threshold?
> Ref: https://gist.github.com/RyanRamchandar/64c5863838940ec67f03
>
appreciate your persistence.. it is working :) will take a look thanks!

**Comment by codefromthecrypt** [comment]
moved okhttp related discussion to a new issue as this topic is tangentially related https://github.com/openzipkin/zipkin/issues/1760

**Comment by dan-tr** [comment]
Issue resolved
